internal/pkg/walker.go

```diff
@@ -0,0 +1,148 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pkg
+
+import (
+	"bytes"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/GoogleContainerTools/kpt/internal/errors"
+	kptfilev1alpha2 "github.com/GoogleContainerTools/kpt/pkg/api/kptfile/v1alpha2"
+	"github.com/monochromegane/go-gitignore"
+	"sigs.k8s.io/kustomize/kyaml/copyutil"
+)
+
+// AllFileMatchFunc is a matcher func that matches all files.
+var AllFileMatchFunc FileMatchFunc = func(string, os.FileInfo) bool {
+	return true
+}
+
+// YamlFileMatchFunc is a matcher func that matches all files with .yaml or
+// .yml extensions.
+var YamlFileMatchFunc FileMatchFunc = func(_ string, info os.FileInfo) bool {
+	if info.IsDir() {
+		return true
+	}
+
+	ext := filepath.Ext(info.Name())
+	if ext == ".yaml" || ext == ".yml" {
+		return true
+	}
+	return false
+}
+
+// KptfileYamlFileMatchFunc is a matcher func that matches the Kptfile and files
+// with either .yaml or .yml extensions.
+var KptfileYamlFileMatchFunc FileMatchFunc = func(s string, info os.FileInfo) bool {
+	if info.IsDir() {
+		return true
+	}
+
+	if info.Name() == kptfilev1alpha2.KptFileName {
+		return true
+	}
+
+	ext := filepath.Ext(info.Name())
+	if ext == ".yaml" || ext == ".yml" {
+		return true
+	}
+	return false
+}
+
+// WalkFunc defines the type for the callback function used by the
+// Walker.Walk function.
+type WalkFunc func(string, os.FileInfo, error) error
+
+// FileMatchFunc is the type of a function that can be provided to the
+// Walker to filter the files that will be provided to the caller.
+type FileMatchFunc func(string, os.FileInfo) bool
+
+// Walker provides functionality for walking the content of a
+// kpt package. It allows for customization of the files and directories
+// that will be provided to the caller through the callback function.
+type Walker struct {
+	// FileMatchFunc defines the function that will be used by the walker to
+	// filter which files should be provided to the callback. By default, all
+	// files will be provided.
+	FileMatchFunc FileMatchFunc
```

Comment:
Recently @Shell32-Natsu introduced a function to skip files in kyaml Readers, have to see how we can use that to implement kptignore aware resource reader.

Reply:
This implementation combines the glob pattern and the skip file function in the upstream kyaml pkg reader into the `FileMatcher` interface. It is harder to include the Ignore pattern here since it requires checking the Kptfile content.
The idea is to have a `KptPackageReader` implementation of the `kio.Reader` interface that we can use in kpt instead of the `kio.LocalPackageReader`.